### PR TITLE
Improve performance of `ListResources`

### PR DESCRIPTION
**Pull request by rosstimothy**
Reduces latency of `proto.AuthService/ListResources` in a variety of ways:

- 6f9ac02: Moves RBAC logging from DEBUG to TRACE 
- c71ef7b: Stores compiled `*regexp.Regexp` in a LRU cache so that they can be reused during RBAC
- ec0860c: Adds a `GetLabel(key string) (value string, ok bool)` to `types.ResourcesWithLabels` to prevent copying the entire label set when we just need to look up keys
- 78d0bef: Prevents loading an extra page from the cache to determine the next key in `auth.ServerWithRoles.ListResources`
- 81b08ed: Modifies `services.UnmarshalServer` to unmarshal directly into a `types.ServerV2` instead of first into a `types.ResourceHeader` to check that the version is `types.V2`

Comparison of `BenchmarkListNodes` from https://github.com/gravitational/teleport/pull/23534/commits/b1715a5d3bb804428efbf9b16def5778c4e4ee5f to https://github.com/gravitational/teleport/pull/23534/commits/ec0860c72047666b533a174a6ea2e1931a87be8f:
```
benchstat b1715a5.txt ec0860c.txt
goos: darwin
goarch: amd64
pkg: github.com/gravitational/teleport/lib/auth
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
             │ b1715a5.txt │             ec0860c.txt             │
             │   sec/op    │   sec/op     vs base                │
ListNodes-16   19.307 ± 5%   1.033 ± 14%  -94.65% (p=0.000 n=10)

             │  b1715a5.txt   │             ec0860c.txt              │
             │      B/op      │     B/op      vs base                │
ListNodes-16   11154.5Mi ± 0%   493.0Mi ± 0%  -95.58% (p=0.000 n=10)

             │  b1715a5.txt  │             ec0860c.txt             │
             │   allocs/op   │  allocs/op   vs base                │
ListNodes-16   112.067M ± 0%   8.341M ± 0%  -92.56% (p=0.000 n=10)
```

### CODE REVIEW

**Review by fspmarshall — COMMENTED** [review]
nit: the old `CombineLabels` strategy would perform the combination s.t. command labels took precedence over static labels (i.e. if a command label and a static label exist for the same key, only the command label would be observed).  As implemented, these `GetLabel` methods are giving precedence to static labels.  I actually prefer this strategy, but it would technically be a breaking change in RBAC behavior, so probably best to change it.

**Review by zmb3 — APPROVED** [review]
nice work!

File: lib/auth/auth_with_roles_test.go [hunk 1]
```diff
@@ -1144,6 +1148,107 @@ func TestSessionRecordingConfigRBAC(t *testing.T) {
...
+
+	ctx := context.Background()
+	srv := newTestTLSServer(b)
+
+	var values []string
+	for i := 0; i < roleCount; i++ {
+		values = append(values, uuid.New().String())
+	}
+
+	values[0] = "evil"
```
Comment (strideynet) [inline lib/auth/auth_with_roles_test.go:1179]: Nit: I know this is just a benchmark but think we could probably use something a bit clearer than "evil" here. Perhaps "hidden" or "inaccessible" ?

**Review by strideynet — APPROVED** [review]
Great work, awesome to see benchmarks being used to ensure improvements are worthwhile.

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**